src/api/tokenList/TokenListApiImpl.ts

```diff
@@ -15,29 +11,47 @@ function _getImageUrl(tokenAddress?: string): string | undefined {
 export class TokenListApiImpl implements TokenList {
   public networkIds: number[]
   private _tokensByNetwork: { [networkId: number]: TokenDetails[] }
+  private _exchangeApi: ExchangeApi
 
-  public constructor(networkIds: number[]) {
+  public constructor(networkIds: number[], exchangeApi: ExchangeApi) {
     this.networkIds = networkIds
+    this._exchangeApi = exchangeApi
 
     // Init the tokens by network
     this._tokensByNetwork = {}
     networkIds.forEach(networkId => {
-      // Get list of tokens
-      const tokens = getTokensByNetwork(networkId).map(token => {
-        // Add image to the tokens
-        return {
-          ...token,
-          image: _getImageUrl(token.addressMainnet),
-        }
-      })
-
-      this._tokensByNetwork[networkId] = tokens
+      // initial value
+      this._tokensByNetwork[networkId] = getTokensByNetwork(networkId)
     })
   }
 
   public getTokens(networkId: number): TokenDetails[] {
     return this._tokensByNetwork[networkId] || []
   }
+
+  public async updateTokenIdsForNetwork(networkId: number): Promise<void> {
+    // on every call fetch initial list again as tokens previously discarded might have been added to exchange
+    const tokens = getTokensByNetwork(networkId)
+    if (!tokens) {
+      return
+    }
+
+    const tokenPromises = tokens.map(token => this.injectExchangeIdIntoToken(token))
+
+    // Resolve promises and remove tokens which are not registered in the exchange
+    this._tokensByNetwork[networkId] = (await Promise.all(tokenPromises)).filter(Boolean) as TokenDetails[]
+  }
+
+  private async injectExchangeIdIntoToken(token: TokenDetails): Promise<TokenDetails | null> {
+    try {
+      // Should throw when address not registered
+      const exchangeId = await this._exchangeApi.getTokenIdByAddress(token.address)
+      return { ...token, id: exchangeId }
+    } catch (e) {
+      log('Token not registered on the exchange %s', token.address, e)
+      return null
```

Comment:
At first I thought this was not right, cause we could return the token details but without the ID.

But on a second thought, maybe is fine to do this for now, since I cannot came up with an use case we have in the app where we have a token without an ID. 
This behaviour would become a problem if we do a interface for adding tokens to the exchange.

Alternatively we do the filtering in the UI. But again, I don't foresee the need in the short-middle term

Reply:
In case we are adding tokens from the UI, we would have the ID once it's added.
So I don't see a scenario where we would have a token in the UI without the ID. Either it's registered in the contract or it's added to it by the user. Never empty.